README.md
# Laravel API Helpers

[![Latest Stable Version](https://poser.pugx.org/fintech-systems/laravel-api-helpers/v/stable?format=flat-square)](https://packagist.org/packages/fintech-systems/laravel-api-helpers)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/fintech-systems/laravel-api-helpers/run-tests?label=tests)](https://github.com/fintech-systems/laravel-api-helpers/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/fintech-systems/laravel-api-helpers/Check%20&%20fix%20styling?label=code%20style)](https://github.com/fintech-systems/laravel-api-helpers/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/fintech-systems/laravel-api-helpers.svg?style=flat-square)](https://packagist.org/packages/fintech-systems/laravel-api-helpers)
[![GitHub license](https://img.shields.io/github/license/fintech-systems/laravel-api-helpers)](https://github.com/fintech-systems/laravel-api-helpers/blob/main/LICENSE.md)

---
## A helper library to facilitate working with APIs

### API: Get, Post, Delete

`public function get(string $url, array $header = [])`

`public function post(string $url, string|array $postFields, array $header = [])`

`public function delete(string $url, string|array $postFields = '', array $header = [])`

Laravel API Helper Command - facilitates caching JSON API responses

Example:

```php
class UserCommand extends LaravelApiHelpersCommand
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'api:get-users {--cached}';

    ...

    /**
     * Execute the console command.
     *
     * @return int
     */
    public function handle()
    {
        if ($file = $this->checkCachedFileExists()) {
            $this->info('A cached API result file was returned');

            return $file;
        }

        $result = file_put_contents($this->cachedFile, Api::getUsers());
        $this->info('The API command was successful');

        return $result;
    }
```

---

## Installation

You can install the package via composer:

```bash
composer require fintech-systems/laravel-api-helpers
```
## Usage

### WhatsApp to WHMCS phone number conversion

```php
$api = new Api();
$result = $api->convertWhatsAppNumberToWhmcsPhoneNumber('27823096710');
expect($result)->toEqual('+27.82 309 6710');

$api = new Api();
$result = $api->convertWhatsAppNumberToWhmcsPhoneNumber('14085551234');
expect($result)->toEqual('+1.408-555-1234');
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [Eugene van der Merwe](https://github.com/fintech-systems)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-api-helpers` will be documented in this file.

## v0.1.5 - 2022-08-19

- had to come upgrade because Yodlee PHP API needs this weird dependency!
- upgraded Orchestra testbench to be L9 compatible (version 7)
- had to remove all packages, upgrade Orchestra, and then add all packages back
- updated composer.json because url is not laravel-api-helpers but just api-helpers
- tests failing because no slack API token available on hand but that's fine because 4/6 are passing

## v0.1.3 - 2021-10-26

- allow union type string|array in post and delete
- attempt at determining if post type is array then http_build_query() but didn't work
- make delete and put same signature
- update readme to show that same signature is sued

## v0.1.1 - 2021-10-25

- revert previous version

## v0.1.0 - 2021-10-25

- breaking change POST now accepts and array for postFields instead of a string

## v0.0.3 - 2021-10-23

- Add a phone number converter for changing from WhatsApp format to WHMCS format
- Add tests for phone number conversion

## 0.0.2 - 2021-10-16

- Add generic CURL DELETE

## 0.0.1 - 2021-10-15

- initial release
- Generic CURL GET & PUT
- LaravelApiHelpersCommand to help with caching


tests/Pest.php
<?php

use FintechSystems\LaravelApiHelpers\Tests\TestCase;

uses(TestCase::class)->in(__DIR__);


tests/ApiTest.php
<?php

use FintechSystems\LaravelApiHelpers\Api;

it('can test', function () {
    expect(true)->toBeTrue();
});

it('can do http get to a jokes api website', function () {
    $api = new Api();
    $result = json_decode($api->get("https://v2.jokeapi.dev/joke/Any"));
    expect($result->error)->toBeFalse();
});

it('can convert 14085551234 to +1.408-555-1234', function () {
    $api = new Api();
    $result = $api->convertWhatsAppNumberToWhmcsPhoneNumber('14085551234');
    expect($result)->toEqual('+1.408-555-1234');
});

it('can convert 27823096710 to +27.82 309 6710', function () {
    $api = new Api();
    $result = $api->convertWhatsAppNumberToWhmcsPhoneNumber('27823096710');
    expect($result)->toEqual('+27.82 309 6710');
});

it('can post an Array to a Slack', function () {
    require 'vendor/autoload.php';

    $dotenv = Dotenv\Dotenv::createImmutable(__DIR__ . "/..");
    $dotenv->load();

    $data = [
        "token" => $_ENV['SLACK_TEST_TOKEN'],
        "channel" => $_ENV['SLACK_TEST_CHANNEL'],
        "text" => "Test 1",
        "username" => "tester",
    ];

    $api = new Api();

    $result = $api->post(
        'https://slack.com/api/chat.postMessage',
        $data
    );

    expect(json_decode($result, true)["ok"])->toEqual(true);
});

it('can post a String to a Slack', function () {
    require 'vendor/autoload.php';

    $dotenv = Dotenv\Dotenv::createImmutable(__DIR__ . "/..");
    $dotenv->load();

    $header = [
        "Content-Type: application/json;charset=utf-8",
        "Authorization: Bearer {$_ENV['SLACK_TEST_TOKEN']}",
    ];

    $data = "{'channel': '" . $_ENV['SLACK_TEST_CHANNEL'] . "',"
            . "'text': 'Test 2',"
            . "'username': 'tester'}";

    $api = new Api();

    $result = $api->post(
        'https://slack.com/api/chat.postMessage',
        $data,
        $header
    );

    expect(json_decode($result, true)["ok"])->toEqual(true);
});


tests/TestCase.php
<?php

namespace FintechSystems\LaravelApiHelpers\Tests;

use FintechSystems\LaravelApiHelpers\LaravelApiHelpersServiceProvider;
use Illuminate\Database\Eloquent\Factories\Factory;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    protected function setUp(): void
    {
        parent::setUp();

        Factory::guessFactoryNamesUsing(
            fn (string $modelName) => 'FintechSystems\\LaravelApiHelpers\\Database\\Factories\\'.class_basename($modelName).'Factory'
        );
    }

    protected function getPackageProviders($app)
    {
        return [
            LaravelApiHelpersServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        config()->set('database.default', 'testing');

        /*
        $migration = include __DIR__.'/../database/migrations/create_laravel-api-helpers_table.php.stub';
        $migration->up();
        */
    }
}


src/Api.php
<?php

namespace FintechSystems\LaravelApiHelpers;

/**
 * An opinioned CURL class the does delete, get, post
 */
class Api
{
    public function get(string $url, array $header = [])
    {
        $curl = curl_init();

        curl_setopt_array($curl, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => '',
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => 'GET',
            CURLOPT_HTTPHEADER => $header,
        ]);

        $response = curl_exec($curl);

        curl_close($curl);

        return $response;
    }

    public function post(string $url, string|array $postFields, array $header = [])
    {
        // gettype($postFields == 'array') ?: $postFields = http_build_query($postFields);

        $curl = curl_init();

        curl_setopt_array($curl, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => '',
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => 'POST',
            CURLOPT_POSTFIELDS => $postFields,
            CURLOPT_HTTPHEADER => $header,
        ]);

        $response = curl_exec($curl);

        curl_close($curl);

        return $response;
    }

    public function delete(string $url, string|array $postFields = '', array $header = [])
    {
        $curl = curl_init();

        curl_setopt_array($curl, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_ENCODING => '',
            CURLOPT_MAXREDIRS => 10,
            CURLOPT_TIMEOUT => 0,
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
            CURLOPT_CUSTOMREQUEST => 'DELETE',
            CURLOPT_POSTFIELDS => $postFields,
            CURLOPT_HTTPHEADER => $header,
        ]);

        $response = curl_exec($curl);

        $statusCode = curl_getinfo($curl, CURLINFO_HTTP_CODE);

        ray($statusCode);

        curl_close($curl);

        return $response;
    }

    /**
     * Convert a WhatsApp Number to WHMCS's Phone Number format
     *
     * Convert 14085551234 to +1.408-555-1234
     * Convert 27823096710 to +27.82 309 6710
     *
     */
    public function convertWhatsAppNumberToWhmcsPhoneNumber($phoneNumber)
    {
        if (substr($phoneNumber, 0, 1) == "1") {
            $countryCode = "+" . substr($phoneNumber, 0, 1) . ".";

            $first = substr($phoneNumber, 1, 3) . "-";
            $second = substr($phoneNumber, 4, 3) . "-";
            $last = substr($phoneNumber, 7, 4);
        } else {
            $countryCode = "+" . substr($phoneNumber, 0, 2) . ".";

            $first = substr($phoneNumber, 2, 2) . " ";
            $second = substr($phoneNumber, 4, 3) . " ";
            $last = substr($phoneNumber, 7, 4);
        }

        return $countryCode . $first . $second . $last;
    }
}
